protofsm/state_machine.go

```diff
@@ -0,0 +1,428 @@
+	case *SendMsgEvent[Event]:
+		sendAndCleanUp := func() error {
+			err := s.daemon.SendMessages(
+				daemonEvent.TargetPeer, daemonEvent.Msgs,
+			)
+			if err != nil {
+				return fmt.Errorf("unable to send msgs: %w", err)
+			}
+
+			// If a post-send event was specified, then we'll
+			// funnel that back into the main state machine now as
+			// well.
+			daemonEvent.PostSendEvent.WhenSome(func(event Event) {
+				s.wg.Add(1)
+				go func() {
+					defer s.wg.Done()
+
+					s.SendEvent(event)
+				}()
+			})
+
+			return nil
+		}
+
+		// If this doesn't have a SendWhen predicate, then we can just
+		// send it off right away.
+		if !daemonEvent.SendWhen.IsSome() {
+			if err := sendAndCleanUp(); err != nil {
+				return err
+			}
+		}
+
+		// Otherwise, this has a SendWhen predicate, so we'll need
+		// launch a goroutine to poll the SendWhen, then send only once
+		// the predicate is true.
+		s.wg.Add(1)
+		go func() {
+			defer s.wg.Done()
+
+			predicateTicker := time.NewTicker(pollInterval)
+			defer predicateTicker.Stop()
+
+			for {
+				select {
+				case <-predicateTicker.C:
+					canSend := fn.MapOptionZero(
+						daemonEvent.SendWhen,
+						func(pred SendPredicate) bool {
+							return pred()
+						},
+					)
+
+					if canSend {
+						sendAndCleanUp()
+						return
+					}
+
+				case <-s.quit:
+					return
+				}
+			}
+		}()
+
+		return nil
+
+	// If this is a disable channel event, then we'll disable the channel.
+	// This is usually done for things like co-op closes.
+	case *DisableChannelEvent:
+		err := s.daemon.DisableChannel(daemonEvent.ChanPoint)
+		if err != nil {
+			return fmt.Errorf("unable to disable channel: %w", err)
+		}
+
+		return nil
+
+	// If this is a broadcast transaction event, then we'll broadcast with
+	// the label attached.
+	case *BroadcastTxn:
+		err := s.daemon.BroadcastTransaction(
+			daemonEvent.Tx, daemonEvent.Label,
+		)
+		if err != nil {
+			// TODO(roasbeef): hook has channel read event event is
+			// hit?
+			return fmt.Errorf("unable to broadcast txn: %w", err)
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("unknown daemon event: %T", event)
+}
+
+// applyEvents applies a new event to the state machine. This will continue
+// until no further events are emitted by the state machine. Along the way,
+// we'll also ensure to execute any daemon events that are emitted.
+func (s *StateMachine[Event, Env]) applyEvents(newEvent Event) (State[Event, Env], error) {
+	// TODO(roasbeef): make starting state as part of env?
+	currentState := s.currentState
+
+	eventQueue := fn.NewQueue(newEvent)
+
+	// Given the next event to handle, we'll process the event, then add
+	// any new emitted internal events to our event queue. This continues
+	// until we reach a terminal state, or we run out of internal events to
+	// process.
+	for nextEvent := eventQueue.Dequeue(); nextEvent.IsSome(); nextEvent = eventQueue.Dequeue() {
+		err := fn.MapOptionZero(nextEvent, func(event Event) error {
```

Comment:
I think instead of using `MapOptionZero`, we should properly check `nextEvent` against nil. IMO this method is a bit of cheating - it leaves out a possible unexpected use case and gives a default zero value, which may cause undefined behavior. For instance, in this case, we'd check against `nextEvent`, and do something(return an error, logging, default to a value, etc) about it if it's empty. Or maybe I misunderstood things here.

Reply:
So a few layers here, but no nil checks are needed at all. The way `Dequeue()` works is that it'll return `None` if there's nothing left in the queue. In that case, `MapOptionZero` will just return `nil`, then the next time round `nextEvent.IsSome()` returns false and the loop exits. Can maybe try to break things up a bit to make easier to understand, but `MapOptionZero` just takes something that returns an error, and calls that if `Some`, otherwise, it returns the zero value of the closure return, which is just `nil`.